### PR TITLE
Fixes line breaks and tabs in serializing errors.

### PR DESCRIPTION
**Pull request by nimakarimipour**
At this moment, NullAway will produce the following output when serializing reporting errors:

```java
foo(X.m()
      ?  X.m2()
    : null)
foo (Object bar) { }
```
```
PASS_NULLABLE   passing @Nullable parameter 'X.m()
            ?                                      X.m2()
            : null' where @NonNull is required   com.uber.foo.Bar    baz()  

```

Since columns are separated by `\t` in `tsv` files and every row should be in one single line, the output should have been the below:
```
PASS_NULLABLE   passing @Nullable parameter 'X.m()\n\t\t? X.m2()\n\t: null' where @NonNull is required   com.uber.foo.Bar  baz()  
```

This PR fixes this issue with a test to verify the correctness.

In addition to `\n` and `\t`, this PR encodes other escape characters (`\r`, `\f` and `\b`) as well.

Regards,

Nima

### CODE REVIEW

**Comment by nimakarimipour** [comment]
Hello @msridhar @lazaroclapp, Would you please tell me what is the latest decision about this. We have two alternatives:
1. Find a way to escape `\n` and `\t` in the error messages.
2. Remove the error message and only serialize `Error Type`.

Please let me know which one you prefer and I will update this PR. At this moment, since annotator can work with only `Error Type` I think removing the error message might be the better solution and add error message when needed in future.

**Comment by msridhar** [comment]
@nimakarimipour I am ok with just serializing `Error Type`, and adding new error types if you need them to distinguish cases for the annotator.  I think that's a better design than trying to match error types based on the message.  That said, if there is any possibility of some other field containing newlines or tabs, we should still fix the encoding there.  But I think that shouldn't be possible?

@lazaroclapp any thoughts?

**Comment by nimakarimipour** [comment]
Hi @msridhar @lazaroclapp. I looked up for convention for lossless `tsv` conversions and it seems the most common is to replace character `\n` with string `"\n"` instead and other conventions including enclosing in quotes are ambiguous. This PR currently follows the first approach. Please let me know if you prefer another way.

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/Serializer.java [hunk 1]
```diff
@@ -54,6 +64,13 @@ public Serializer(FixSerializationConfig config) {
     initializeOutputFiles(config);
   }
 
+  public static String encodeEscapeSequences(String content) {
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/fixserialization/Serializer.java:67]: Why not just put this method in `ErrorInfo` (along with the `escapes` map), since that is the only place it is used?  That way both could be private to that class.

Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/Serializer.java:67]: Also it'd be good to add a bit of Javadoc to this method indicating what is being escaped (maybe with a link to a web page describing the escaping)

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java [hunk 2]
```diff
@@ -828,6 +828,41 @@ public void errorSerializationTest() {
...
+            "   public void run() {",
+            "     // BUG: Diagnostic contains: passing @Nullable parameter 'm.hashCode()",
+            "     foo(m.hashCode() == 2 || m.toString().equals('\\t') ? \t\n\t\n new Object() : null);",
+            "   }",
+            "   public void foo(Object o) { }",
+            "}")
+        .setExpectedOutputs(
+            new ErrorDisplay(
+                "PASS_NULLABLE",
+                "passing @Nullable parameter 'm.hashCode() == 2 || m.toString().equals('\\t') ? \\t\\n\\t\\n new Object() : null'",
```
Comment (msridhar) [inline nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java:858]: 👍

**Review by msridhar — CHANGES_REQUESTED** [review]
Looks good, just a minor thing.  Probably @lazaroclapp should also take a quick look

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/Serializer.java (hunk 1 above)
Comment (nimakarimipour) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/Serializer.java]: 707889b4cdf57fa473a4a0eb4fa71cbfc5b2d59a

**Review by msridhar — APPROVED** [review]
LGTM!

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java [hunk 3]
```diff
@@ -22,21 +22,53 @@
...
+   *
+   * @param str String to process.
+   * @return returns modified str which its special characters are escaped.
+   */
+  private static String escapeSpecialCharacters(String str) {
+    // escape existing backslashes
+    str = str.replaceAll(Pattern.quote("\\"), Matcher.quoteReplacement("\\\\"));
+    // escape special characters
+    for (Character key : escapes.keySet()) {
+      str = str.replaceAll(String.valueOf(key), "\\\\" + escapes.get(key));
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java:67]: Why four slashes here? Shouldn't we just want to add a single `\` to the actual internal string representation, meaning the encoding on the string literally should just be `\\`? (Above we are re-encoding single slashes as double slashes, but we want e.g. new line replaced by `\n` internally: single-slash, followed by n)

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java [hunk 4]
```diff
@@ -828,6 +828,41 @@ public void errorSerializationTest() {
...
+                "-XepOpt:NullAway:SerializeFixMetadata=true",
+                "-XepOpt:NullAway:FixSerializationConfigPath=" + configPath))
+        .addSourceLines(
+            "com/uber/Test.java",
+            "package com.uber;",
+            "public class Test {",
+            "   Object m = new Object();",
+            "   public void run() {",
+            "     // BUG: Diagnostic contains: passing @Nullable parameter 'm.hashCode()",
+            "     foo(m.hashCode() == 2 || m.toString().equals('\\t') ? \t\n\t\n new Object() : null);",
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java:851]: Can we make the new lines here explicit in the test case? E.g. instead of passing a test case string with `\n`, just put the new lines inside a new comma-separated string in its own line:

```
" foo(m.hashCode() == 2 || m.toString().equals('\\t') ? \t",
"\t",
" new Object() : null);",
```
?

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java (hunk 3 above)
Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java:67]: It seems to be working based on the test, so maybe I am missing something here. But, in that case we want a comment explaining it.

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java [hunk 5]
```diff
@@ -828,6 +828,41 @@ public void errorSerializationTest() {
...
+            "}")
+        .setExpectedOutputs(
+            new ErrorDisplay(
+                "PASS_NULLABLE",
+                "passing @Nullable parameter 'm.hashCode() == 2 || m.toString().equals('\\\\t') ? \\t\\n\\t\\n new Object() : null'",
+                "com.uber.Test",
+                "run()"))
+        .setFactory(errorDisplayFactory)
+        .setOutputFileNameAndHeader(ERROR_FILE_NAME, ERROR_FILE_HEADER)
+        .doTest();
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java:863]: Actually... is this test even testing the TSV serialization code? It seem to me that this is just comparing the `ErrorDisplay` objects themselves, but doesn't go through the process of serializing/deserializing them :/

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java (hunk 4 above)
Comment (nimakarimipour) [reply nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java]: b57eef578749b860ee06e1f79e59319a8c3fcf6a

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java (hunk 3 above)
Comment (nimakarimipour) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java:69]: In regular expression `\` is a separate special character from `\n` and `\t` and etc. To match exactly `\` we need to pass "\\\\". However in order to create `\\` with string in java, each backslash must also be escaped, therefore, we need four `\` to make regex match only the character `\`.

24acf2781088486f4f62ef0133d59f7503735101

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java (hunk 5 above)
Comment (nimakarimipour) [reply nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java:865]: Would you please tell me if it is necessary to do everything in reverse and construct the actual message  to pass the test ?

Currently we are writing the output below in the file:
```
passing @Nullable parameter 'm.hashCode() == 2 || m.toString().equals('\\t') ? \t\n\t\n new Object() : null' where @NonNull is required
```

Which is exactly what we need to reconstruct the actual message, if we don't escape special characters, the output below will be written to the file and the test will fail:
```
passing @Nullable parameter 'm.hashCode() == 2 || m.toString().equals('\t') ? 	
	
 new Object() : null' where @NonNull is required	
 ```

Isn't this sufficient to test escaping special characters in serialization ?

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java (hunk 3 above)
Comment (nimakarimipour) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java]: I used `quoteReplacement()` for both cases to make it more readable.

944d6ac53fcc19715c41e6f884aa5ac152fdc142

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java [hunk 6]
```diff
@@ -22,21 +22,55 @@
...
+   * Conventions for lossless conversion to TSV</a>
+   *
+   * @param str String to process.
+   * @return returns modified str which its special characters are escaped.
+   */
+  private static String escapeSpecialCharacters(String str) {
+    // regex needs "\\" to match character '\', each must also be escaped in string to create "\\",
+    // therefore we need four "\".
+    // escape existing backslashes
+    str = str.replaceAll(Pattern.quote("\\"), Matcher.quoteReplacement("\\\\"));
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java:66]: @nimakarimipour this line means you will replace any occurrence of `\\` with `\\\\`, right?  Why do we need to do this?

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java (hunk 3 above)
Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java]: My point is that I don't think the second argument of `replaceAll(...)` is a regex, but a string. In which case "\\\\n" becomes the string for [`\`,`\`,`n`] in memory, where we want [`\`,`n`]. Unless I am missing something.

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java (hunk 5 above)
Comment (lazaroclapp) [reply nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java:865]: Ok. Never mind. This seems correct to me. I was missing the fact that `SerializationTestHelper.readActualOutputs` is actually parsing the strings back from the TSV. That means it's going through the TSV serialization code, so this is what you want. My bad!

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java (hunk 6 above)
Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java:66]: Ok, I think I understand this and this is correct, though very annoying:

1. The string "\\" in the Java source code here represents in-memory string literal (as `char[]`): [`\`]
2. Similarly, "\\\\" in the source is the literal [`\`,`\`]
3. `Patern.quote("\\")` does the replacement, producing [`\`] -> [`\`,`\`], but this is needed because the first argument of `String.replaceAll(...)` is a regex, which means that literal string matches the string [`\`] (otherwise, a single `\` in the string literal would be interpreted as part of the regex language, waiting for a escaped character inside the regex, leading to an incomplete regex)
4. So far, so good for my understanding of Java regexes. But what I missed the first time around is that, while the second parameter to `String.replaceAll(...)` is **not** a regex, it is something called a "replacement string", with its own special syntax. In particular, the docs have this to say:

> Note that backslashes (\) and dollar signs ($) in the replacement string may cause the results to be different than if it were being treated as a literal replacement string. Dollar signs may be treated as references to captured subsequences as described above, and backslashes are used to escape literal characters in the replacement string.

See [docs](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/regex/Matcher.html#replaceAll(java.lang.String)) for `Matcher.replaceAll`, which is the underlying implementation of `String.replaceAll`.

Thus, for the second argument here: Java source code string `"\\\\"` becomes literal [`\`,`\`], which gets expanded by `Matcher.quoteReplacement` into [`\`,`\`,`\`,`\`], which finally gets interpreted by `String.replaceAll(...)` as a "replacement string" replacing every match by the literal string [`\`,`\`].

Basically this: https://xkcd.com/1638/

But the code looks correct to me :)

**Review by lazaroclapp — APPROVED** [review]

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java (hunk 6 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java:66]: 🤣 great explanation.  Now I'm going to try to forget this stuff exists 🙂

Comment (nimakarimipour) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java:66]: @msridhar This actually replaces each occurrence of `\` with `\\` in text. And in order to show `\` in java string, we have to show it with `\\`. The reason for that is that we want the actual `\t` be present in the serialization and not get mixed with the `\t` we inserted ourself. 
<pre>
The output ".equals('\t')     == true" will be serialized to ".equals('\\t')\t == true"
</pre>
